pkg/controller/resourceclaim/controller_test.go

```diff
@@ -0,0 +1,335 @@
+			name:            "stay-reserved",
+			pods:            []*v1.Pod{testPodWithResource},
+			key:             claimKey(testClaimReserved),
+			claims:          []*v1.ResourceClaim{testClaimReserved},
+			expectedClaims:  []v1.ResourceClaim{*testClaimReserved},
+			expectedMetrics: expectedMetrics{0, 0},
+		},
+		{
+			name:            "clear-reserved",
+			pods:            []*v1.Pod{},
+			key:             claimKey(testClaimReserved),
+			claims:          []*v1.ResourceClaim{testClaimReserved},
+			expectedClaims:  []v1.ResourceClaim{*testClaim},
+			expectedMetrics: expectedMetrics{0, 0},
+		},
+		{
+			name:            "remove-reserved",
+			pods:            []*v1.Pod{testPod},
+			key:             claimKey(testClaimReservedTwice),
+			claims:          []*v1.ResourceClaim{testClaimReservedTwice},
+			expectedClaims:  []v1.ResourceClaim{*testClaimReserved},
+			expectedMetrics: expectedMetrics{0, 0},
+		},
+	}
+
+	for _, tc := range tests {
+		// Run sequentially because of global logging and global metrics.
+		t.Run(tc.name, func(t *testing.T) {
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			var objects []runtime.Object
+			for _, pod := range tc.pods {
+				objects = append(objects, pod)
+			}
+			for _, claim := range tc.claims {
+				objects = append(objects, claim)
+			}
+
+			fakeKubeClient := createTestClient(objects...)
+			if tc.expectedMetrics.numFailures > 0 {
+				fakeKubeClient.PrependReactor("create", "resourceclaims", func(action k8stesting.Action) (handled bool, ret runtime.Object, err error) {
+					return true, nil, apierrors.NewConflict(action.GetResource().GroupResource(), "fake name", errors.New("fake conflict"))
+				})
+			}
+			setupMetrics()
+			informerFactory := informers.NewSharedInformerFactory(fakeKubeClient, controller.NoResyncPeriodFunc())
+			podInformer := informerFactory.Core().V1().Pods()
+			claimInformer := informerFactory.Core().V1().ResourceClaims()
+
+			c, err := NewController(fakeKubeClient, podInformer, claimInformer)
+			if err != nil {
+				t.Fatalf("error creating ephemeral controller : %v", err)
+			}
+			ec, _ := c.(*resourceClaimController)
+
+			// Ensure informers are up-to-date.
+			go informerFactory.Start(ctx.Done())
+			defer func() {
+				cancel()
+				informerFactory.Shutdown()
+			}()
+			informerFactory.WaitForCacheSync(ctx.Done())
+			cache.WaitForCacheSync(ctx.Done(), podInformer.Informer().HasSynced, claimInformer.Informer().HasSynced)
+
+			err = ec.syncHandler(context.TODO(), tc.key)
+			if err != nil && !tc.expectedError {
+				t.Fatalf("unexpected error while running handler: %v", err)
+			}
+			if err == nil && tc.expectedError {
+				t.Fatalf("unexpected success")
+			}
+
+			claims, err := fakeKubeClient.CoreV1().ResourceClaims("").List(ctx, metav1.ListOptions{})
+			if err != nil {
+				t.Fatalf("unexpected error while listing claims: %v", err)
+			}
+			assert.Equal(t, normalizeClaims(tc.expectedClaims), normalizeClaims(claims.Items))
```

Comment:
Assert continues. I think that's the right thing to do here because there are other checks that can still run.

assert.Equal does the same as manually calling cmp.Diff with less code.